Detekt

mikepenz opened the pull request:
* add detekt to project
* enable detekt on travis

-------

Alternative to spotless. Replacing https://github.com/mikepenz/FastAdapter/pull/785

AllanWang commented on build.gradle:
```diff
@@ -10,7 +10,7 @@ buildscript {
                 buildTools: "29.0.0",
                 minSdk    : 14,
-                targetSdk : 29
+                targetSdk : 28
```
Why was this reverted to 28?

mikepenz replied:
roboelectric does not support 29 yet :/

AllanWang commented on build.gradle:
```diff
@@ -34,12 +34,16 @@ buildscript {
         classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:${versions.kotlin}"
         classpath 'com.github.dcendents:android-maven-gradle-plugin:2.1'
         classpath 'com.jfrog.bintray.gradle:gradle-bintray-plugin:1.8.4'
+        classpath "io.gitlab.arturbosch.detekt:detekt-gradle-plugin:1.0.0-RC15"
```
Might want to put version within `ext.versions` since it occurs in 3 places

mikepenz replied:
yes true. I will do that.

AllanWang commented:
Does the formatter actually work? The detekt command itself just seems to label some warnings. It might take `detektIdeaFormat` to fix it though

mikepenz commented:
I tried locally with wrong indents and it failed the build, so it should fail the travis build too when the wrong indents are used for example. 

the idea stuff requires IDEA with the format.sh script which is not available on travis 

it will not modify anything

AllanWang commented:
Weird. For #788 it doesn't fail to build, even if I move some indents around


```
47 kotlin files were analyzed.

Ruleset: comments
Ruleset: complexity
Ruleset: empty-blocks
Ruleset: exceptions
Ruleset: formatting - 30min debt
	NoBlankLineBeforeRbrace - [PsiWhiteSpace] at /...SelectExtension.kt:163:64
	NoBlankLineBeforeRbrace - [PsiWhiteSpace] at /...SelectExtension.kt:167:62
	NoBlankLineBeforeRbrace - [PsiWhiteSpace] at /...AbstractItem.kt:60:42
	NoBlankLineBeforeRbrace - [PsiWhiteSpace] at /...AbstractItem.kt:69:45
	NoBlankLineBeforeRbrace - [PsiWhiteSpace] at /...AbstractItem.kt:78:46
	NoBlankLineBeforeRbrace - [PsiWhiteSpace] at /...OnBindViewHolderListenerImpl.kt:72:14
Ruleset: naming - 5min debt
	MatchingDeclarationName - [IInterceptor.kt] at /...IInterceptor.kt:1:1
Ruleset: performance
Ruleset: potential-bugs
Ruleset: style

Overall debt: 35min

Complexity Report:
	- 4317 lines of code (loc)
	- 2297 source lines of code (sloc)
	- 1595 logical lines of code (lloc)
	- 1453 comment lines of code (cloc)
	- 488 McCabe complexity (mcc)
	- 7 number of total code smells
	- 63 % comment source ratio
	- 305 mcc per 1000 lloc
	- 4 code smells per 1000 lloc

Project Statistics:
	- number of properties: 193
	- number of functions: 282
	- number of classes: 47
	- number of packages: 7
	- number of kt files: 47

Successfully generated HTML report at /...
Successfully generated Checkstyle XML report at /...

detekt finished in 1635 ms.
```

mikepenz commented:
@AllanWang I checked. The problem was that we were within the 10 warnings threshold per module which was ok. 

I set it now to 0 and fixed all the remaining warnings and adjusted others
